### PR TITLE
Search API for books

### PR DESCRIPTION
**Pull request by Lynoxe**
I implemented the search API where you can enter a text query and it returns a list of books corresponding to the query. It is fetched from https://openlibrary.org. I only implemented the requests of the following shape: [https://openlibrary.org/search.json?q={query}](https://openlibrary.org/search.json?q=the+lord+of+the+rings)

I tested  (with a mock web server) the api directly, using it through a remote data source and using it through a repository, it all works fine. Though, more tests would be useful as I only checked a simple 2 documents case. I didn't commit that, but I also tested that every level of the API works online using the real openLibrary database. I didn't commit those tests because they were very informal and they might have not passed the CI, later on we should create a good `Screen` to display the results and create android tests for this `Screen`.

For now I decided to let `SearchResult` and `Document` at the `domain` layer because I didn't write any conversion, so I just assumed that until we don't have a way to convert these to the correct meaningful data structure, we might use them directly in the project.

I strictly followed the example we all agreed to base our project on.

### CODE REVIEW

File: domain/src/main/java/com/github/leuludyha/domain/util/FileReader.kt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.github.leuludyha.domain.util
+
+import java.io.IOException
+import java.io.InputStreamReader
+
+object FileReader {
```
Comment (eugenefle) [inline domain/src/main/java/com/github/leuludyha/domain/util/FileReader.kt:6]: It seems this FileReader is only being used to test data sources, why is it in the domain package ?

File: domain/src/main/java/com/github/leuludyha/domain/util/Result.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.github.leuludyha.domain.util
+
+import java.io.Serializable
+
+sealed class Result<T>(
```
Comment (eugenefle) [inline domain/src/main/java/com/github/leuludyha/domain/util/Result.kt:5]: It seems to me that  the Result data class is part of domain model, why create a new util package ?

File: app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/openLibrary/OpenLibraryViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import dagger.hilt.android.lifecycle.HiltViewModel
+
+import com.github.leuludyha.domain.model.Search
+import com.github.leuludyha.domain.useCase.SearchUseCase
+import com.github.leuludyha.domain.util.Result
+
+
+@HiltViewModel
+class OpenLibraryViewModel @Inject constructor(
```
Comment (eugenefle) [inline app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/openLibrary/OpenLibraryViewModel.kt:19]: Again I don't think a ViewModel has anything to do with the OpenLibrary API.

**Review by eugenefle — CHANGES_REQUESTED** [review]
The actual content looks good to me, except for some details on data sources return types. However, the naming and placing of some interfaces seems a little confusing, I have added suggestions for those.

File: data/src/main/java/com/github/leuludyha/data/api/SearchApi.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.github.leuludyha.data.api
+
+import com.github.leuludyha.domain.model.Search
+import retrofit2.Response
+import retrofit2.http.GET
+import retrofit2.http.Query
+
+interface SearchApi {
```
[the remark this answers is not in the document]
Comment (Lynoxe) [reply data/src/main/java/com/github/leuludyha/data/api/SearchApi.kt:8]: For this one, I disagree as my idea was to expose multiple APIs, following the same logic as the openlibrary website (one API for search, one API for books, one for covers, etc.)
So I thought doing the generalisation to a global Library interface at either the data source or the repository layer

File: app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/openLibrary/OpenLibraryViewModel.kt (hunk 3 above)
Comment (Lynoxe) [reply app/src/main/java/com/github/leuludyha/ibdb/presentation/screen/openLibrary/OpenLibraryViewModel.kt:19]: I could have chosen a better name but `OpenLibrary` here doesn't refer to the website but to a possible `Screen` that we would create later. However we probably wouldn't name it `OpenLibrary`. But in any case a `ViewModel` is inherently linked to a specific `View` (`Screen` in our architecture)

File: data/src/main/java/com/github/leuludyha/data/api/SearchApi.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (eugenefle) [reply data/src/main/java/com/github/leuludyha/data/api/SearchApi.kt:8]: Oh yes that's true, I remember now

**Review by eugenefle — APPROVED** [review]
Everything looks good now

**Pull request merged**